src/charm.py

```diff
@@ -236,6 +239,31 @@ def is_kubeconfig_available(self):
                     return True
         return False
 
+    def set_service_cidr(self, event):
+        """Repeat received service-cidr from k8s-cp to each kube-ovn unit.
+
+        service-cidr is received over the cni relation only from
+        kubernetes-control-plane units.  the kube-ovn peer relation
+        shares the value around to each kube-ovn unit.
+        """
+        cni_service_cidr = event.relation.data[event.unit].get("service-cidr")
+        if cni_service_cidr:
+            for relation in self.model.relations["kube-ovn"]:
```

Comment:
I'm a little worried that Juju hooks could fire in this order:

* install
* config-changed
* cni-relation-joined
* cni-relation-changed
* kube-ovn-joined

We would arrive here during cni-relation-changed, but before the kube-ovn relation has been joined. Will the relation exist under `self.model.relations["kube-ovn"]` at that time? I'll see if I can find out.

Reply:
Interesting. It looks like the relation appears even before the first kube-ovn-created hook. I put in a simple debug print:

```
log.info('GKK: %s %s' % (os.environ.get('JUJU_HOOK_NAME'), self.model.relations["kube-ovn"]))
```

and saw:

```
unit-kube-ovn-0: 13:41:40 INFO unit.kube-ovn/0.juju-log GKK: install []
unit-kube-ovn-0: 13:41:45 INFO unit.kube-ovn/0.juju-log cni:13: GKK: cni-relation-created [<ops.model.Relation kube-ovn:1>]
unit-kube-ovn-0: 13:41:50 INFO unit.kube-ovn/0.juju-log kube-ovn:1: GKK: kube-ovn-relation-created [<ops.model.Relation kube-ovn:1>]
unit-kube-ovn-0: 13:41:50 INFO unit.kube-ovn/0.juju-log GKK: leader-elected [<ops.model.Relation kube-ovn:1>]
unit-kube-ovn-0: 13:41:51 INFO unit.kube-ovn/0.juju-log GKK: config-changed [<ops.model.Relation kube-ovn:1>]
unit-kube-ovn-0: 13:41:55 INFO unit.kube-ovn/0.juju-log GKK: start [<ops.model.Relation kube-ovn:1>]
unit-kube-ovn-0: 13:41:57 INFO unit.kube-ovn/0.juju-log kube-ovn:1: GKK: kube-ovn-relation-joined [<ops.model.Relation kube-ovn:1>]
unit-kube-ovn-0: 13:41:58 INFO unit.kube-ovn/0.juju-log kube-ovn:1: GKK: kube-ovn-relation-changed [<ops.model.Relation kube-ovn:1>]
unit-kube-ovn-0: 13:41:59 INFO unit.kube-ovn/0.juju-log cni:13: GKK: cni-relation-joined [<ops.model.Relation kube-ovn:1>]
unit-kube-ovn-0: 13:42:00 INFO unit.kube-ovn/0.juju-log cni:13: GKK: cni-relation-changed [<ops.model.Relation kube-ovn:1>]
```

So I think we're good?

Reply:
Hmm, I'm not seeing it. Can you walk me through how that can occur?

Reply:
Starting situation:
```
k8s-cp/0           active idle
   kube-ovn/0      active idle
k8s-worker/0       active idle
   kube-ovn/1      active idle
```

`kube-ovn/0` gets the `service-cidr` from `k8s-cp/0`, tells all his "current" peers. the only current peer is `kube-ovn/1`

`kube-ovn/2` joins the peer relationship later on

the joined hook for `kube-ovn` relation isn't defined
the changed hook for `kube-ovn` relation doesn't share the `service-cidr` with the newly joined unit

only a change of the cni relation to an existing `kube-ovn` unit will actually update the all the `kube-ovn` peer relation data.  This leaves all new `kube-ovn` units without the `service-cidr`

Reply:
I think that case should be okay. If you look at how we set the data on the relation:

```python
for relation in self.model.relations["kube-ovn"]:
    relation.data[self.unit]["service-cidr"] = cni_service_cidr
```

Remote unit IDs don't appear in that code, and this translates pretty much directly to the Juju relation-set calls that happen behind the scenes. We're not so much sending it to current peers as we are sending it to Juju to put in the kube-ovn:1 relation data for this unit. Later units that join should still be able to see that data, even if they weren't there at the time it was set.

Here's what I see when trying this. Setting foo=bar on the kube-ovn/1 unit's kube-ovn relation:

```
$ juju run --unit kube-ovn/1 -- relation-ids kube-ovn
kube-ovn:1
$ juju run --unit kube-ovn/1 -- relation-set -r kube-ovn:1 foo=bar
```

kube-ovn/0 can get it:
```
$ juju run --unit kube-ovn/0 -- relation-get -r kube-ovn:1 foo kube-ovn/1
bar
```

Adding a new unit:
```
juju add-unit kubernetes-worker
```

Some minutes later, this leads to the creation of kube-ovn/2, which can also get the data that was set previously:
```
$ juju run --unit kube-ovn/2 -- relation-ids kube-ovn
kube-ovn:1

$ juju run --unit kube-ovn/2 -- relation-get -r kube-ovn:1 foo kube-ovn/1
bar
```

Am I missing something?

Reply:
Ahh good, you cleared it up for me.  The fact they're all using the same relation-id is what i think was throwing me.  I was misunderstanding that each new unit was going to show up on a uniq relation-id.

Reply:
I think it's usually one relation per application that's related to the endpoint. So yeah, what you're worried about could very well be a problem for a `requires` or `provides` relation where applications can be added or removed. I think what saves us here is that the peers relation is implicit and fixed so we only ever get one of it.

I could see setting `service-cidr` on the `kube-ovn-relation-joined` hook or something like that regardless. Seems like good practice for handling relations in general, and it might give us some peace of mind that there isn't some weird corner case we're missing.

Reply:
Yeah

I'm not going to make that change of handling `kube-ovn-relation-joined` and copying my `get_service_cidr()` into my unit's published data on the relation

This would impede future changes of the `service_cidr` during a cidr expansion.

when the cni relation updates one kube-ovn, that unit would change the kube-ovn relation causing all the other kube-ovn units to disagree as to what the `service-cidr` should be.  ONLY `kube-ovn` units with CNI relations containing the original source of truth should ever update the kube-ovn relation data.